File: libs/ml-kernel/src/models/feature.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use uuid::Uuid;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FeatureSample {
    pub entity_key: String,
    pub value: Value,
    pub observed_at: Option<DateTime<Utc>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FeatureDefinition {
    pub id: Uuid,
    pub name: String,
    pub entity_name: String,
    pub data_type: String,
    pub description: String,
    pub status: String,
    pub offline_source: String,
    pub transformation: String,
    pub online_enabled: bool,
    pub online_namespace: String,
    pub batch_schedule: String,
    pub freshness_sla_minutes: i32,
    pub tags: Vec<String>,
    pub samples: Vec<FeatureSample>,
    pub last_materialized_at: Option<DateTime<Utc>>,
    pub last_online_sync_at: Option<DateTime<Utc>>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ListFeaturesResponse {
    pub data: Vec<FeatureDefinition>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CreateFeatureRequest {
    pub name: String,
    pub entity_name: String,
    pub data_type: String,
    #[serde(default)]
    pub description: String,
    #[serde(default)]
    pub offline_source: String,
    #[serde(default)]
    pub transformation: String,
    #[serde(default)]
    pub online_enabled: bool,
    #[serde(default)]
    pub online_namespace: String,
    #[serde(default = "default_batch_schedule")]
    pub batch_schedule: String,
    #[serde(default = "default_freshness_sla")]
    pub freshness_sla_minutes: i32,
    #[serde(default)]
    pub tags: Vec<String>,
    #[serde(default)]
    pub samples: Vec<FeatureSample>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct UpdateFeatureRequest {
    pub name: Option<String>,
    pub entity_name: Option<String>,
    pub data_type: Option<String>,
    pub description: Option<String>,
    pub status: Option<String>,
    pub offline_source: Option<String>,
    pub transformation: Option<String>,
    pub online_enabled: Option<bool>,
    pub online_namespace: Option<String>,
    pub batch_schedule: Option<String>,
    pub freshness_sla_minutes: Option<i32>,
    pub tags: Option<Vec<String>>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct MaterializeFeatureRequest {
    #[serde(default)]
    pub samples: Vec<FeatureSample>,
    pub mode: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OnlineFeatureSnapshot {
    pub feature_id: Uuid,
    pub namespace: String,
    pub source: String,
    pub values: Vec<FeatureSample>,
    pub fetched_at: DateTime<Utc>,
}

fn default_batch_schedule() -> String {
    "0 * * * *".to_string()
}

fn default_freshness_sla() -> i32 {
    60
}

File: libs/ml-kernel-go/models/feature.go
package models

import (
	"encoding/json"
	"time"

	"github.com/google/uuid"
)

// FeatureSample is a single (entity_key, value) observation.
type FeatureSample struct {
	EntityKey  string          `json:"entity_key"`
	Value      json.RawMessage `json:"value"`
	ObservedAt *time.Time      `json:"observed_at"`
}

// FeatureDefinition is the catalog row.
type FeatureDefinition struct {
	ID                   uuid.UUID       `json:"id"`
	Name                 string          `json:"name"`
	EntityName           string          `json:"entity_name"`
	DataType             string          `json:"data_type"`
	Description          string          `json:"description"`
	Status               string          `json:"status"`
	OfflineSource        string          `json:"offline_source"`
	Transformation       string          `json:"transformation"`
	OnlineEnabled        bool            `json:"online_enabled"`
	OnlineNamespace      string          `json:"online_namespace"`
	BatchSchedule        string          `json:"batch_schedule"`
	FreshnessSLAMinutes  int32           `json:"freshness_sla_minutes"`
	Tags                 []string        `json:"tags"`
	Samples              []FeatureSample `json:"samples"`
	LastMaterializedAt   *time.Time      `json:"last_materialized_at"`
	LastOnlineSyncAt     *time.Time      `json:"last_online_sync_at"`
	CreatedAt            time.Time       `json:"created_at"`
	UpdatedAt            time.Time       `json:"updated_at"`
}

type ListFeaturesResponse struct {
	Data []FeatureDefinition `json:"data"`
}

// CreateFeatureRequest defaults: batch_schedule="0 * * * *",
// freshness_sla_minutes=60.
type CreateFeatureRequest struct {
	Name                string          `json:"name"`
	EntityName          string          `json:"entity_name"`
	DataType            string          `json:"data_type"`
	Description         string          `json:"description,omitempty"`
	OfflineSource       string          `json:"offline_source,omitempty"`
	Transformation      string          `json:"transformation,omitempty"`
	OnlineEnabled       bool            `json:"online_enabled,omitempty"`
	OnlineNamespace     string          `json:"online_namespace,omitempty"`
	BatchSchedule       string          `json:"batch_schedule,omitempty"`
	FreshnessSLAMinutes int32           `json:"freshness_sla_minutes,omitempty"`
	Tags                []string        `json:"tags,omitempty"`
	Samples             []FeatureSample `json:"samples,omitempty"`
}

type UpdateFeatureRequest struct {
	Name                *string   `json:"name,omitempty"`
	EntityName          *string   `json:"entity_name,omitempty"`
	DataType            *string   `json:"data_type,omitempty"`
	Description         *string   `json:"description,omitempty"`
	Status              *string   `json:"status,omitempty"`
	OfflineSource       *string   `json:"offline_source,omitempty"`
	Transformation      *string   `json:"transformation,omitempty"`
	OnlineEnabled       *bool     `json:"online_enabled,omitempty"`
	OnlineNamespace     *string   `json:"online_namespace,omitempty"`
	BatchSchedule       *string   `json:"batch_schedule,omitempty"`
	FreshnessSLAMinutes *int32    `json:"freshness_sla_minutes,omitempty"`
	Tags                *[]string `json:"tags,omitempty"`
}

type MaterializeFeatureRequest struct {
	Samples []FeatureSample `json:"samples,omitempty"`
	Mode    *string         `json:"mode,omitempty"`
}

type OnlineFeatureSnapshot struct {
	FeatureID uuid.UUID       `json:"feature_id"`
	Namespace string          `json:"namespace"`
	Source    string          `json:"source"`
	Values    []FeatureSample `json:"values"`
	FetchedAt time.Time       `json:"fetched_at"`
}

const (
	DefaultBatchSchedule       = "0 * * * *"
	DefaultFreshnessSLAMinutes int32 = 60
)
